### PR TITLE
Generic TopK implementation

### PR DESCRIPTION
**Pull request by killeent**
This PR makes TopK generic. The main changes required to do this were:

1. Extend the float --> unsigned integer bijective function to all other Tensor types
2. Make it possible to act on data types whose underlying representation requires 64bits

### CODE REVIEW

**Comment by killeent** [comment]
@soumith @wickedfoo This is ready.

File: lib/THC/THCTensorTopK.cuh [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+    return v - 128;
+  }
+};
+
+template <>
+struct TopKTypeConfig<short> {
+  typedef unsigned int RadixType;
+
+  static inline __device__ RadixType convert(short v) {
+    return 32768u + v;
```
Comment (wickedfoo) [inline lib/THC/THCTensorTopK.cuh]: add a static assert that sizeof(short) == 2 to ensure that this is a correct constant

File: lib/THC/THCTensorTopK.cuh [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+    return v - 32768;
+  }
+};
+
+template <>
+struct TopKTypeConfig<int> {
+  typedef unsigned int RadixType;
+
+  static inline __device__ RadixType convert(int v) {
+    return 2147483648u + v;
```
Comment (wickedfoo) [inline lib/THC/THCTensorTopK.cuh]: add a static assert that sizeof(int) == 4 to ensure that this is a correct constant

File: lib/THC/THCTensorTopK.cuh [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+    return v - 2147483648u;
+  }
+};
+
+template <>
+struct TopKTypeConfig<long> {
+  typedef unsigned long long int RadixType;
+
+  static inline __device__ RadixType convert(long v) {
+    return 9223372036854775808ull + v;
```
Comment (wickedfoo) [inline lib/THC/THCTensorTopK.cuh]: add a static assert that sizeof(long) == 8 to ensure that this is a correct constant

**Pull request merged**